### ISSUE TITLE
IOptions<T> should support getting named options

### ISSUE THREAD

**Issue by jez9999** [dotnet/runtime#84478]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Shouldn't `IOptions<T>` support getting named options?  I know you can use `IOptionsSnapshot` and `IOptionsMonitor` but they monitor the updating of the JSON config file which presumably comes with a performance hit.  If it's appropriate for some classes to use the one-off binding of `IOptions<T>`, why isn't it appropriate for a singleton to use it for one-off binding when it wants to get named options?  In fact, I may actually want the behaviour of my singleton only changing its configured behaviour when the service is restarted rather than live as the file is edited (that seems like somewhat surprising behaviour as I'm used to settings taking effect only on restart).

### Describe the solution you'd like

Add something like `.Get(string optionsName)` to `IOptions<T>`.

**Comment by tarekgh** [comment]
@jez9999 Can you provide more details about your scenario and how incorporating `IOptions<TOptions>.Get(optionName)` will be beneficial? Additionally, could you clarify why using `IOptionsMonitor<TOptions>.Get(String)` isn't sufficient for your needs? Providing some code examples to demonstrate what you are trying to accomplish would also be useful.

**Comment by jez9999** [comment]
@tarekgh If I have a singleton service that, for example, persists data to a database every 10 seconds, and I update the config file, I might expect the updated settings to take effect upon restart, not immediately.  `IOptionsMonitor` gets the update immediately.  It could even lead to inconsistent behaviour if code relies on getting a complete group of settings on startup.

`IOptionsMonitor` may be "sufficient", but then why does `IOptions` exist at all?  Isn't it meant to be a simpler way to get a snapshot of options that were configured when the program started?  What I'm saying is, if you're doing that for some settings, it's only consistent to do it for all settings, even ones used by singletons.

**Comment by tarekgh** [comment]
@jez9999 

> but then why does IOptions exist at all?

It looks to me there is some misunderstanding here. `IOptions` provides strongly typed access to groups of related settings. .NET exposes other options interfaces to support different scenarios. like `IOptionsSnapshot` and `IOptionsMonitor`. You may look at the [Options interfaces doc](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-7.0#options-interfaces) for more info. As stated in the doc, `IOptions` only read the configurations once at the startup and never read the configuration again. Isn't that the behavior you are requesting here?

**Comment by jez9999** [comment]
But it doesn't support named options.  What if you want named options with the "just read the configuration once at startup" behaviour?

**Comment by ScarletKuro** [comment]
> But it doesn't support named options. What if you want named options with the "just read the configuration once at startup" behaviour?

This still doesn't answer the question why you need the `.Get(string optionsName)` method in `IOptions`. `IOptions` is added only once at startup using code similar to the following:
```CSharp
services.Configure<DbOptions>(Configuration.GetSection("DbOptions"));
```
Afterward, you can inject it via Dependency Injection using `IOptions<DbOptions>`.
If you updated the configuration, it will apply only after the application restart.

If you still want to access an IOptions instance by name, you can inject IConfiguration and use the following code:
```CSharp
var section = Configuration.GetSection("DbOptions").Get<DbOptions>();
IOptions<DbOptions> option = Microsoft.Extensions.Options.Options.Create(section);
```

**Comment by jez9999** [comment]
That `IConfiguration` code looks horrible, it's the sort of thing that should only go in the initial DI setup.  The consumer should be able to do something tidy like `IOptions<DbOptions> foo; ... foo.Get("QA")`.  If the `IConfiguration` injection is appropriate, why does .NET provide `IOptions` via DI at all?

**Comment by davidfowl** [comment]
It’s unclear what this issue I asking for now that it’s been answered. There are 3 interfaces for consuming strongly typed configuration listed in the docs. There’s also INamed* equivalents of those interfaces.

What are you trying to solve?
What are you doing today?
What are you trying to improve?

**Comment by jez9999** [comment]
@davidfowl I don't see that it has been answered.  How do I consume named options with `IOptions`?  The other interfaces do something different, keeping the latest version of the JSON file monitored, whereas `IOptions` doesn't do that.

**Comment by davidfowl** [comment]
When using named options, the only way to get the value is via IOptionsMonitor or IOptionsSnapshot. The reason IOptions\<T\> doesn’t have names is because they were added later. We could add to the existing interface but it wouldn’t be usable in net standard 2.0.

**Comment by jez9999** [comment]
Indeed.  And adding it's what I'm suggesting.

**Comment by rafal-zieba** [comment]
I's been more than a year now. If you don't want to add support for named options to `IOptions<T>` interface than maybe we could get one more interface that behaves like `IOptions<T>` (no monitoring) but has the support. `IOptionsWithNamedOptionsSupoport<T>` for lack of a better name. Without it named options are far less useful. We need to accept performance hit in e.g. every web request (monitoring).

**Comment by davidfowl** [comment]
We can do this now that we have default interface methods, but it would work on .NET 8+, not down level. Still, I don't really understand why you can't use `IOptionsMonitor<T>.Get(name)`.

**Comment by julealgon** [comment]
@jez9999 if you can change your code ever so slightly to inject the options model instead of grabbing the named instance, something like this should work:

```csharp
services
    .AddSingletonKeyed<IOptions<MyOptions>>("QA", p => p.GetRequiredService<IOptionsMonitor>().Get("QA"));
```

Then you can inject this into your service either using the attribute:
```csharp
public class MyService([FromKeyedServices("QA")]IOptions<MyOptions> options)
...
```

Or just grab the value directly at method level:
```csharp
public class MyService(IServiceProvider provider)
{
    public void MyMethod(string environment)
    {
        var options = provider.GetRequiredKeyedService<IOptions<MyOptions>>(environment);
    }
}
```

Or even drop the identifier altogether if you can do this:
```csharp
services
    .AddSingletonKeyed<IOptions<MyOptions>>("QA", p => p.GetRequiredService<IOptionsMonitor>().Get("QA"))
    .AddTransient<MyService>(p => ActivatorUtilities.CreateInstance<MyService>(
        p, 
        p.GetRequiredKeyedService<IOptions<MyOptions>>());
```

On a personal level... I still firmly believe the entire `IOptions` abstraction along with named options is a workaround to limitations in the Microsoft container (some of which were lifted when keyed registrations were added), but I end up using it anyways just for consistency's sake as it has become the de-facto standard for these kinds of settings.

**Comment by julealgon** [comment]
@davidfowl 
> We can do this now that we have default interface methods, but it would work on .NET 8+, not down level. Still, I don't really understand why you can't use `IOptionsMonitor<T>.Get(name)`.

`IOptionsMonitor` will force rebinding and re-running the config pipeline for that `T` instance, which is what the OP wants to avoid.

Just using it directly won't fix this unless the class where you are using it is registered as singleton itself.

One of the ways to make it behave like `IOptions` is to override the registration like I mentioned above, which combined with keyed registrations ends up producing the same effect as a singleton named options would.
